File: deep_search-dev/api/crawler_api.py
import asyncio
from playwright.async_api import async_playwright
import logging
from typing import List, Dict, Any


# 设置日志级别
logging.getLogger("playwright").setLevel(logging.ERROR)


class CrawlerAPI:
    def __init__(self, logger=None):
        self.browser = None
        self.context = None
        self.playwright = None
        self.logger = logger or logging.getLogger(__name__)
        
    async def init_browser(self):
        """初始化浏览器"""
        if not self.browser:
            try:
                self.playwright = await async_playwright().start()

                self.browser = await self.playwright.chromium.launch(
                    headless=True,
                    timeout=10000  # 添加浏览器启动超时
                )
                
                if self.browser:
                    self.context = await self.browser.new_context()
                else:
                    self.logger.error("浏览器启动失败")
            except Exception as e:
                self.logger.error(f"浏览器初始化出错: {str(e)}")
                self.browser = None
                self.context = None
                self.playwright = None
                # 重新抛出异常以便上层处理
                raise
            
    async def close_browser(self):
        """关闭浏览器"""
        if self.context:
            await self.context.close()
            self.context = None
        if self.browser:
            await self.browser.close()
            self.browser = None
        if self.playwright:
            await self.playwright.stop()
            self.playwright = None
            
    async def crawl_single_url(self, url: str) -> Dict[str, Any]:
        """
        爬取单个URL的内容
        
        Args:
            url (str): 要爬取的URL
            
        Returns:
            Dict[str, Any]: 包含url和content的字典
        """
        if not self.context:
            await self.init_browser()
            
        # 再次检查context是否初始化成功
        if not self.context:
            return {
                "url": url,
                "content": "浏览器初始化失败"
            }
            
        try:
            page = await self.context.new_page()
            # 设置更短的超时时间和更宽松的等待条件
            await page.goto(url, wait_until="domcontentloaded", timeout=15000)
            
            # 等待一小段时间确保页面加载
            await asyncio.sleep(1)
            
            # 获取页面文本内容（不包含HTML标签）
            content = await page.inner_text("body")
            
            # 关闭页面
            await page.close()
            
            return {
                "url": url,
                "content": content if content else "无法获取页面文本内容"
            }
        except asyncio.TimeoutError:
            logging.error(f"爬取URL {url} 超时")
            return {
                "url": url,
                "content": "爬取超时，可能是网站加载太慢或无法访问"
            }
        except Exception as e:
            logging.error(f"爬取URL {url} 时出错: {str(e)}")
            return {
                "url": url,
                "content": f"爬取失败: {str(e)}"
            }
            
    async def crawl_urls(self, urls: List[str]) -> List[Dict[str, Any]]:
        """
        并行爬取多个URL的内容
        
        Args:
            urls (List[str]): 要爬取的URL列表
            
        Returns:
            List[Dict[str, Any]]: 包含每个URL的url和content的字典列表
        """
        if not self.context:
            await self.init_browser()
            
        # 创建任务列表
        tasks = [self.crawl_single_url(url) for url in urls]
        
        # 并行执行所有任务
        results = await asyncio.gather(*tasks, return_exceptions=True)
        
        # 处理结果
        formatted_results = []
        for result in results:
            if isinstance(result, Exception):
                logging.error(f"爬取任务出错: {str(result)}")
                continue
            formatted_results.append(result)
            
        return formatted_results
        
    async def __aenter__(self):
        """异步上下文管理器入口"""
        await self.init_browser()
        return self
        
    async def __aexit__(self, exc_type, exc_val, exc_tb):
        """异步上下文管理器出口"""
        await self.close_browser()

File: Agent/deep_search-dev-go/api/crawler_api.go
package api

import (
	"context"
	"fmt"
	"sync"
	"time"

	"github.com/playwright-community/playwright-go"
	"deepsearch/logs"
)

// CrawlResult 表示爬取结果
type CrawlResult struct {
	URL     string `json:"url"`
	Content string `json:"content"`
}

// CrawlerAPI 封装了网页爬取功能
type CrawlerAPI struct {
	browser playwright.Browser
	logger  *logs.Logger
	mu      sync.Mutex
}

// NewCrawlerAPI 创建一个新的 CrawlerAPI 实例
func NewCrawlerAPI(logger *logs.Logger) *CrawlerAPI {
	return &CrawlerAPI{
		logger: logger,
	}
}

// InitBrowser 初始化浏览器
func (c *CrawlerAPI) InitBrowser() error {
	c.mu.Lock()
	defer c.mu.Unlock()
	
	if c.browser != nil {
		return nil
	}
	
	c.logger.Info("正在初始化浏览器...")
	
	// 启动Playwright
	pw, err := playwright.Run()
	if err != nil {
		return fmt.Errorf("启动Playwright失败: %v", err)
	}
	
	// 启动浏览器
	browser, err := pw.Chromium.Launch(playwright.BrowserTypeLaunchOptions{
		Headless: playwright.Bool(true),
	})
	if err != nil {
		return fmt.Errorf("启动浏览器失败: %v", err)
	}
	
	c.browser = browser
	c.logger.Info("浏览器初始化成功")
	
	return nil
}

// CloseBrowser 关闭浏览器
func (c *CrawlerAPI) CloseBrowser() error {
	c.mu.Lock()
	defer c.mu.Unlock()
	
	if c.browser == nil {
		return nil
	}
	
	c.logger.Info("正在关闭浏览器...")
	
	err := c.browser.Close()
	c.browser = nil
	
	if err != nil {
		c.logger.Errorf("关闭浏览器失败: %v", err)
		return err
	}
	
	c.logger.Info("浏览器已关闭")
	return nil
}

// CrawlSingleURL 爬取单个URL的内容
func (c *CrawlerAPI) CrawlSingleURL(ctx context.Context, url string) (*CrawlResult, error) {
	// 确保浏览器已初始化
	if c.browser == nil {
		if err := c.InitBrowser(); err != nil {
			return &CrawlResult{
				URL:     url,
				Content: "浏览器初始化失败",
			}, err
		}
	}
	
	// 创建页面
	page, err := c.browser.NewPage()
	if err != nil {
		return &CrawlResult{
			URL:     url,
			Content: fmt.Sprintf("创建页面失败: %v", err),
		}, err
	}
	defer page.Close()
	
	// 设置超时
	ctx, cancel := context.WithTimeout(ctx, 30*time.Second)
	defer cancel()
	
	// 导航到URL
	_, err = page.Goto(url)
	if err != nil {
		return &CrawlResult{
			URL:     url,
			Content: "页面加载超时，可能是网站加载太慢或无法访问",
		}, fmt.Errorf("页面加载超时: %v", err)
	}
	
	// 等待页面加载完成
	page.WaitForTimeout(3000)
	
	// 等待一小段时间确保页面加载完成
	time.Sleep(1 * time.Second)
	
	// 获取页面文本内容
	textContent, err := page.InnerText("body")
	if err != nil {
		return &CrawlResult{
			URL:     url,
			Content: "无法获取页面文本内容",
		}, err
	}
	
	if textContent == "" {
		textContent = "无法获取页面文本内容"
	}
	
	return &CrawlResult{
		URL:     url,
		Content: textContent,
	}, nil
}

// CrawlURLs 并行爬取多个URL的内容
func (c *CrawlerAPI) CrawlURLs(ctx context.Context, urls []string) ([]*CrawlResult, error) {
	// 确保浏览器已初始化
	if c.browser == nil {
		if err := c.InitBrowser(); err != nil {
			return nil, err
		}
	}
	
	c.logger.Infof("开始爬取 %d 个URL...", len(urls))
	
	// 创建结果通道
	resultChan := make(chan *CrawlResult, len(urls))
	errorChan := make(chan error, len(urls))
	
	// 使用 WaitGroup 等待所有 goroutine 完成
	var wg sync.WaitGroup
	wg.Add(len(urls))
	
	// 为每个 URL 启动一个 goroutine 进行爬取
	for _, url := range urls {
		go func(u string) {
			defer wg.Done()
			
			// 检查上下文是否已取消
			select {
			case <-ctx.Done():
				resultChan <- &CrawlResult{
					URL:     u,
					Content: "爬取已取消",
				}
				return
			default:
			}
			
			// 爬取单个 URL
			result, err := c.CrawlSingleURL(ctx, u)
			if err != nil {
				c.logger.Errorf("爬取URL %s 时出错: %v", u, err)
				errorChan <- err
				return
			}
			
			resultChan <- result
		}(url)
	}
	
	// 等待所有 goroutine 完成
	go func() {
		wg.Wait()
		close(resultChan)
		close(errorChan)
	}()
	
	// 收集结果
	var results []*CrawlResult
	for result := range resultChan {
		results = append(results, result)
	}
	
	// 检查是否有错误
	for err := range errorChan {
		if err != nil {
			c.logger.Errorf("爬取任务出错: %v", err)
		}
	}
	
	c.logger.Infof("爬取完成，共获取到 %d 个结果", len(results))
	return results, nil
}